### PR TITLE
[Arch, Eval] Allowing injecting additional dependency to OD runtime docker image

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Part of #2404, #3230.

For some evaluation (e.g., Logical reasoning, MINT) to work, you have to pre-install dependencies on the "OD's Python interpreter," which is used to power the "IPython."

We allow each eval to use their own `Dockerfile`, so the dependencies can be pre-installed for "bash", but not "IPython," since that's built on top of the user-provided base image on the fly:

https://github.com/OpenDevin/OpenDevin/blob/5469f67cb9a1f1b345dc872a1cdb799451fe5e27/evaluation/logic_reasoning/Dockerfile#L1-L5

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

- Add `od_runtime_extra_deps` to `SandboxConfig`, which allows us to inject dependencies to the `Dockerfile.j2` template when we build an OD runtime image for an arbitrary base image.

Example use case: Install an additional package inside IPython before the start of the evaluation
https://github.com/OpenDevin/OpenDevin/blob/5469f67cb9a1f1b345dc872a1cdb799451fe5e27/evaluation/logic_reasoning/run_infer.py#L44-L64

---
**Other references**

### CODE REVIEW

File: opendevin/core/config.py [hunk 1]
```diff
@@ -173,6 +178,7 @@ class SandboxConfig(metaclass=Singleton):
     use_host_network: bool = False
     initialize_plugins: bool = True
     update_source_code: bool = False
+    od_runtime_extra_deps: str = ''
```
Comment (tobitege) [inline opendevin/core/config.py:181]: Why not also `str | None`? I guess it won't be used too often?

File: opendevin/core/config.py [hunk 2]
```diff
@@ -154,6 +154,11 @@ class SandboxConfig(metaclass=Singleton):
         initialize_plugins: Whether to initialize plugins.
         update_source_code: Whether to update the source code in the EventStreamRuntime.
             Used for development of EventStreamRuntime.
+        od_runtime_extra_deps: The extra dependencies to install in the runtime image (typically used for evaluation).
+            This will be rendered into the end of the Dockerfile that builds the runtime image.
+            It can contains any valid shell commands (e.g., pip install numpy).
```
Comment (tobitege) [inline opendevin/core/config.py:159]: `contain`

File: opendevin/runtime/utils/runtime_build.py [hunk 3]
```diff
@@ -185,6 +196,7 @@ def build_runtime_image(
     docker_client: docker.DockerClient,
     update_source_code: bool = False,
     save_to_local_store: bool = False,  # New parameter to control saving to local store
+    extra_deps: str = '',  # whether to install extra dependencies inside the image
```
Comment (tobitege) [inline opendevin/runtime/utils/runtime_build.py:199]: `optional` not `whether`?
also, why not `str | none`?

**Review by tobitege — APPROVED** [review]
LGTM, just nits

File: opendevin/core/config.py (hunk 1 above)
Comment (xingyaoww) [reply opendevin/core/config.py]: fixed!

File: opendevin/runtime/utils/runtime_build.py (hunk 3 above)
Comment (xingyaoww) [reply opendevin/runtime/utils/runtime_build.py]: updated to `str | none`

**Pull request merged**